crates/core/transaction/src/view.rs

```diff
@@ -294,3 +466,458 @@ impl TryFrom<pbt::MemoPlaintextView> for MemoPlaintextView {
+        let mut sct = tct::Tree::new();
+        for _ in 0..5 {
+            let random_note = Note::generate(&mut OsRng, &test_keys::ADDRESS_0, value);
+            sct.insert(tct::Witness::Keep, random_note.commit())
+                .unwrap();
+        }
+        sct.insert(tct::Witness::Keep, note.commit()).unwrap();
+        sct.insert(tct::Witness::Keep, note2.commit()).unwrap();
+
+        let auth_path = sct.witness(note.commit()).unwrap();
+        let auth_path2 = sct.witness(note2.commit()).unwrap();
+
+        let gm = asset::Cache::with_known_assets().get_unit("gm").unwrap();
+        let gn = asset::Cache::with_known_assets().get_unit("gn").unwrap();
+        let trading_pair = TradingPair::new(gm.id(), gn.id());
+
+        let delta_1 = Amount::from(100_000u64);
+        let delta_2 = Amount::from(0u64);
+        let fee = Fee::default();
+        let claim_address: Address = test_keys::ADDRESS_0.deref().clone();
+        let plaintext = SwapPlaintext::new(
+            &mut OsRng,
+            trading_pair,
+            delta_1,
+            delta_2,
+            fee,
+            claim_address,
+        );
+
+        // Add a single spend and output to the transaction plan such that the
+        // transaction balances.
+        let plan = TransactionPlan {
+            transaction_parameters: TransactionParameters {
+                expiry_height: 0,
+                fee: Fee::default(),
+                chain_id: "".into(),
+            },
+            actions: vec![
+                SpendPlan::new(&mut OsRng, note, auth_path.position()).into(),
+                SpendPlan::new(&mut OsRng, note2, auth_path2.position()).into(),
+                OutputPlan::new(&mut OsRng, value3, test_keys::ADDRESS_1.deref().clone()).into(),
+                SwapPlan::new(&mut OsRng, plaintext.clone()).into(),
+            ],
+            detection_data: Some(DetectionDataPlan {
+                clue_plans: vec![CluePlan::new(
+                    &mut OsRng,
+                    test_keys::ADDRESS_1.deref().clone(),
+                    1.try_into().unwrap(),
+                )],
+            }),
+            memo: None,
+        };
+
+        let transaction_view = TransactionView {
+            anchor: penumbra_tct::Root(Hash::zero()),
+            binding_sig: Signature::from([0u8; 64]),
+            body_view: TransactionBodyView {
+                action_views: plan
+                    .actions
+                    .iter()
+                    .filter_map(|x| {
+                        convert_action(&Cache::with_known_assets(), &test_keys::FULL_VIEWING_KEY, x)
+                    })
+                    .collect(),
+                transaction_parameters: plan.transaction_parameters.clone(),
+                detection_data: None,
+                memo_view: None,
+            },
+        };
+
+        let transaction_summary = TransactionView::summary(&transaction_view);
+
+        assert_eq!(transaction_summary.effects.len(), 2);
```

Comment:
should add assertions about expected required and provided imbalances and sanity check final result.